### PR TITLE
Free products screen

### PR DESCRIPTION
**Pull request by flegald**
Screen to display free games currently on offer from Epic Games.

API endpoint: `https://store-site-backend-static.ak.epicgames.com/freeGamesPromotions` is used to obtain current promotion information

[Git Issue](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/issues/305)

free-heroic

-------------------------------------------------------------
Use the following Checklist if you have changed something on the Backend or Frontend:

- [x] Tested the feature and it's working on a current and clean install.
- [x] Tested the main App features and they are still working on a current and clean install. (Login, Install, Play, Uninstall, Move games, etc.)
- [x] Created / Updated Tests (If necessary)
- [x] Created / Updated documentation (If necessary)

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -44,6 +45,11 @@ function App() {
               <Library library={library} />
             </div>
           </Route>
+          <Route exact path="/free-games">
```
Comment (Nocccer) [inline src/App.tsx:48]: Why not Route with component and also why not move the class "content" into the FreeProduct function ?

```suggestion
          <Route exact path="/free-games" component={FreeProducts} />
```

File: src/screens/FreeProducts/utils/apiResponseType.tsx [hunk 2]
```diff
@@ -0,0 +1,143 @@
+/**
```
Comment (Nocccer) [inline src/screens/FreeProducts/utils/apiResponseType.tsx:1]: Do we really need all this interfaces ? 
Why not fetching the data you need in the backend and return a new interface which contains all information you need ?

**Review by Nocccer — CHANGES_REQUESTED** [review]
Overall looks good. Still have some open suggestions/questions.

File: src/screens/FreeProducts/utils/apiResponseType.tsx (hunk 2 above)
Comment (dragonDScript) [reply src/screens/FreeProducts/utils/apiResponseType.tsx:1]: Why not have all this?

**Review by dragonDScript — CHANGES_REQUESTED** [review]
LGTM, but the frontend needs to be improved. I'd put it in the library.

File: src/screens/FreeProducts/utils/apiResponseType.tsx (hunk 2 above)
Comment (flavioislima) [reply src/screens/FreeProducts/utils/apiResponseType.tsx:1]: I do agree that the ideal would be to do all requests on BE and just send the needed information to the frontend like we are doing with the other requests. This way we can avoid making the frontend too bloated. 👍🏽

**Comment by flavioislima** [comment]
> LGTM, but the frontend needs to be improved. I'd put it in the library. Or in a separate window (popup) that is spawned when the app starts and when the user requests it. We should create a figma document and share our ideas there.

I think we can keep it here for now because since we are going to develop a sidebar, it will be only one more button there. 
But on the library, like a filter, I don't know, could work as well.

File: src/screens/FreeProducts/index.tsx [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+const FreeProducts = () => {
+  const { t, i18n } = useTranslation()
+  let lang = i18n.language
+  if (i18n.language === 'pt') {
+    lang = 'pt-BR'
+  }
+  const [products, setProducts] = useState<Element[]>([])
+
+  const fetchProducts = async () => {
```
Comment (flavioislima) [inline src/screens/FreeProducts/index.tsx:22]: In my opinion, we could improve the `requestFreeProducts` request to return the `validProducts` already, most of this logic would work better on the backend and the frontend should be used only for rendering that response.

**Review by flavioislima — CHANGES_REQUESTED** [review]
I think looks good overall, but the approach of treating the data on the frontend is not the best one I would say. So moving this logic to the BE could be better and then the frontend would just send a request and get the list and all the info needed to open the game page.

What do you think?

File: src/screens/FreeProducts/utils/apiResponseType.tsx (hunk 2 above)
Comment (Nocccer) [reply src/screens/FreeProducts/utils/apiResponseType.tsx:1]: Doing exactly that with the wine ge implementation. Fetching information in the BE and just return the needed information.

**Comment by dragonDScript** [comment]
> I think looks good overall, but the approach of treating the data on the frontend is not the best one I would say. So moving this logic to the BE could be better and then the frontend would just send a request and get the list and all the info needed to open the game page.
> 
> What do you think?

I think that we should split the backend and the frontend. Frontend needs to be refactored. So merge backend quickly and then work on frontend. what do you think?

**Comment by flegald** [comment]
Sounds good. I will refactor the API response formatting to the backend and make a few smaller changes here and there. 
I do not know where exactly to put the page right now @Nocccer, I dont know if the Library is the best place?

File: electron/utils.ts [hunk 4]
```diff
@@ -208,11 +208,18 @@ async function checkCommandVersion(
   return found;
 }
 
+async function fetchFreeProducts() {
```
Comment (Nocccer) [inline electron/utils.ts:211]: Bring over all functions from FreeProducts/utils.tsx and prepare the data here.
The Return should be Element[].

**Comment by dragonDScript** [comment]
I'm getting this error when cloning this branch: 
Unhandled Rejection (Error): Error invoking remote method 'requestFreeProducts': No handler registered for 'requestFreeProducts'

**Pull request closed without merging**